flink-connectors/flink-hcatalog/src/main/scala/org/apache/flink/table/hive/functions/HiveSimpleUDF.scala

```
@@ -0,0 +1,87 @@
+import org.apache.hadoop.hive.ql.exec.{FunctionRegistry, UDF}
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDFUtils.ConversionHelper
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory
+import org.apache.hadoop.hive.serde2.typeinfo.{PrimitiveTypeInfo, TypeInfo, TypeInfoFactory}
+
+import scala.annotation.varargs
+
+/**
+  * A Hive UDF Wrapper which behaves as a Flink-table ScalarFunction.
+  *
+  * This class has to have a method with @varargs annotation. For scala will compile
+  * <code> eval(args: Any*) </code> to <code>eval(args: Seq)</code>.
+  * This will cause an exception in Janino compiler.
+  */
+class HiveSimpleUDF(className: String) extends ScalarFunction {
+
+  @transient
+  private lazy val functionWrapper = HiveFunctionWrapper(className)
+
+  @transient
+  private lazy val function = functionWrapper.createFunction[UDF]()
+
+  @transient
+  private var typeInfos: util.List[TypeInfo] = _
+
+  @transient
+  private var objectInspectors: Array[ObjectInspector] = _
+
+  @transient
+  private var conversionHelper: ConversionHelper = _
+
+  @transient
+  private var method: Method = _
+
+  @varargs
+  def eval(args: AnyRef*) : Any = {
+    if (null == typeInfos) {
+      typeInfos = new util.ArrayList[TypeInfo]()
+      args.foreach(arg => {
+          typeInfos.add(TypeInfoFactory.getPrimitiveTypeInfoFromJavaPrimitive(arg.getClass))
+      })
+      method = function.getResolver.getEvalMethod(typeInfos)
+
+      objectInspectors = new Array[ObjectInspector](typeInfos.size())
+      args.zipWithIndex.foreach { case (_, i) =>
+        objectInspectors(i) = PrimitiveObjectInspectorFactory.getPrimitiveJavaObjectInspector(
+          typeInfos.get(i).asInstanceOf[PrimitiveTypeInfo])
+      }
+      conversionHelper = new ConversionHelper(method, objectInspectors)
+    }
+
+    val mappedArgs = args.map {
```

Maybe it is better to throw an exception because of an unsupported type BigDecimal type, such that the user can manually cast the BigDecimal to double. Otherwise the user is not aware of losing precision.